### PR TITLE
[2.0] Add tests for IPv6 proxy subjectAltNames (forward-port #2419)

### PR DESCRIPTION
**Pull request by sethmlarson**
Brings #2419 to the `main` branch. Notably this changes where the stripping of `[]` characters happens to match what was released in 1.26.7.

Part of https://github.com/urllib3/urllib3/issues/2530

### CODE REVIEW

**Comment by sethmlarson** [comment]
This PR is failing until https://github.com/urllib3/urllib3/pull/2529 is merged.

**Review by jalopezsilva — COMMENTED** [review]
Small nits, change looks good. Let's close #2529 first?

File: test/conftest.py [hunk 1]
```diff
@@ -113,7 +163,70 @@ def no_san_server_with_different_commmon_name(
...
+    ca = trustme.CA()
+    # non localhost common name
+    server_cert = ca.issue_cert("example.com")
+
+    with run_server_in_thread("https", "localhost", tmpdir, ca, server_cert) as cfg:
+        yield cfg
+
+
+@pytest.fixture
+def ipv4_san_proxy(
```
[the remark this answers is not in the document]
Comment (sethmlarson) [reply test/conftest.py]: Done!

**Comment by jalopezsilva** [comment]
Changes look good but there are some assertion failures that look legit:

```
AssertionError: assert 'no appropriate subjectAltName fields were found' in "[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: Hostname mismatch, certificate is not valid for 'localhost'. (_ssl.c:1131)"
```

The error message returned is slightly different from what we were expecting.

**Comment by jalopezsilva** [comment]
Looks like there are still failures?

```
_____________ TestHTTPSProxyVerification.test_https_proxy_ipv6_san _____________
urllib3.exceptions.SSLError: [SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: Hostname mismatch, certificate is not valid for '[::1]'. (_ssl.c:1131)

The above exception was the direct cause of the following exception:

urllib3.exceptions.ProxyError: ('Unable to connect to proxy', SSLError(SSLCertVerificationError(1, "[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: Hostname mismatch, certificate is not valid for '[::1]'. (_ssl.c:1131)")))
```

**Comment by sethmlarson** [comment]
@jalopezsilva I think this one's ready for a review again. Duplicating the stripping logic isn't super great but it's either duplicating the logic (one of which already exists on 1.26.x) or applying the same logic twice on both `assert_hostname` and `server_hostname`. We lose either way :/

Let me know if you think of a better way to handle this.

**Review by jalopezsilva — DISMISSED** [review]
Agreed, the duplication is not ideal but it's so small that I think it's fine.

There's one failure but it seems to be securetransport noise. I think you're officially good to go 🚀

File: test/with_dummyserver/test_https.py [hunk 2]
```diff
@@ -1051,23 +1051,61 @@ def test_hostname_checks_common_name_respected(
...
+            ipv4_san_server.port,
+            cert_reqs="CERT_REQUIRED",
+            ca_certs=ipv4_san_server.ca_certs,
+        ) as https_pool:
+            r = https_pool.request("GET", "/")
+            assert r.status == 200
+
+
+class TestHTTPS_IPv6Addr:
+    @pytest.mark.xfail  # TODO: Remove test in urllib3/urllib3#2532
```
[the remark this answers is not in the document]
Comment (sethmlarson) [reply test/with_dummyserver/test_https.py]: Yeah we don't need that test any more since we don't match IPv6 addresses in commonName.

**Pull request merged**